src/deluge/gui/views/automation_instrument_clip_view.cpp

```diff
@@ -0,0 +1,3138 @@
+			else if (!velocity || isUIModeWithinRange(auditionPadActionUIModes)) {
+				if (sdRoutineLock && !allowSomeUserActionsEvenWhenInCardRoutine) {
+					return ActionResult::REMIND_ME_OUTSIDE_CARD_ROUTINE; // Allowable sometimes if in card routine.
+				}
+				auditionPadAction(velocity, y, Buttons::isShiftButtonPressed());
+			}
+			setDisplayParameterNameTimer();
+		}
+	}
+
+	return ActionResult::DEALT_WITH;
+}
+
+//edit pad action
+//handles single and multi pad presses for automation editing and for parameter selection on the automation overview
+//stores pad presses in the EditPadPresses struct of the instrument clip view
+
+void AutomationInstrumentClipView::editPadAction(bool state, uint8_t yDisplay, uint8_t xDisplay, uint32_t xZoom) {
+
+	InstrumentClip* clip = getCurrentClip();
+	Instrument* instrument = (Instrument*)clip->output;
+
+	char modelStackMemory[MODEL_STACK_MAX_SIZE];
+	ModelStackWithTimelineCounter* modelStack = currentSong->setupModelStackWithCurrentClip(modelStackMemory);
+
+	// If button down
+	if (state) {
+		if (!isSquareDefined(xDisplay)) {
+
+			return;
+		}
+		// If this is a note-length-edit press...
+		//needed for Automation
+		if (clip->lastSelectedParamID != kNoLastSelectedParamID && instrumentClipView.numEditPadPresses == 1
+		    && ((int32_t)(instrumentClipView.timeLastEditPadPress + 80 * 44 - AudioEngine::audioSampleTimer) < 0)) {
+
+			int32_t firstPadX = 255;
+			int32_t firstPadY = 255;
+
+			// Find that original press
+			int32_t i;
+			for (i = 0; i < kEditPadPressBufferSize; i++) {
+				if (instrumentClipView.editPadPresses[i].isActive) {
+
+					firstPadX = instrumentClipView.editPadPresses[i].xDisplay;
+					firstPadY = instrumentClipView.editPadPresses[i].yDisplay;
+
+					break;
+				}
+			}
+
+			if (firstPadX != 255 && firstPadY != 255) {
+				handleMultiPadPress(modelStack, clip, firstPadX, firstPadY, xDisplay, yDisplay);
+			}
+		}
+
+		// Or, if this is a regular create-or-select press...
+		else {
+
+			instrumentClipView.timeLastEditPadPress = AudioEngine::audioSampleTimer;
+			// Find an empty space in the press buffer, if there is one
+			int32_t i;
+			for (i = 0; i < kEditPadPressBufferSize; i++) {
+				if (!instrumentClipView.editPadPresses[i].isActive) {
+					break;
+				}
+			}
+			if (i < kEditPadPressBufferSize) {
+				handleSinglePadPress(modelStack, clip, xDisplay, yDisplay);
+
+				instrumentClipView.shouldIgnoreVerticalScrollKnobActionIfNotAlsoPressedForThisNotePress = false;
+
+				// If this is the first press, record the time
+				if (instrumentClipView.numEditPadPresses == 0) {
+					instrumentClipView.timeFirstEditPadPress = AudioEngine::audioSampleTimer;
+					instrumentClipView.shouldIgnoreHorizontalScrollKnobActionIfNotAlsoPressedForThisNotePress = false;
+				}
+
+				instrumentClipView.editPadPresses[i].isActive = true;
+				instrumentClipView.editPadPresses[i].yDisplay = yDisplay;
+				instrumentClipView.editPadPresses[i].xDisplay = xDisplay;
```

Comment:
we're sure this can't break `instrumentClipView` somehow?

Reply:
re-using the struct you mean? as far as I can tell, no - but without a lot of testing and usage I won't know for sure.

if we want to be safer I could re-define the struct in my view. I was just trying to reduce my footprint as much as possible.

my understanding is that this struct is strictly used to keep track of what pads have been pressed so that notes can be recorded in the sequencer or so that you can detect single/multi pad presses for example.

Reply:
probably fine then